drivers/pwm/pwm_numaker.c

```diff
@@ -0,0 +1,574 @@
+	}
+
+	/* Get CAP int status */
+	cap_intsts = (epwm->CAPIF & cap_int_mask);
+
+	/* PWM counter is timing critical, to avoid print msg from irq_isr
+	 *  until getting capture cycles.
+	 */
+	LOG_DBG("===[%s] Channel=0x%x, CAPIEN=0x%x, CAPIF=0x%x, capIntMask=0x%x", __func__,
+		st_channel, epwm->CAPIEN, epwm->CAPIF, cap_int_mask);
+	if (cap_intsts != 0x00) { /* Capture Interrupt */
+		/* Clear CAP int status */
+		epwm->CAPIF = cap_intsts;
+
+		/* Rising latch or Falling latch */
+		for (i = st_channel; i <= end_channel; i++) {
+			capture = &data->capture[i];
+			if (capture == NULL) {
+				continue;
+			}
+			if (((EPWM_CAPTURE_INT_RISING_LATCH << i) |
+			     (EPWM_CAPTURE_INT_FALLING_LATCH << i)) &
+			    cap_intsts) {
+
+				EPWM_DisableCaptureInt(epwm, i,
+						       EPWM_CAPTURE_INT_RISING_LATCH |
+							       EPWM_CAPTURE_INT_FALLING_LATCH);
+				/* Calculate cycles */
+				status = __pwm_numaker_get_cap_cycle(
+					epwm, i, data->capture[i].curr_edge_mode,
+					data->capture[i].next_edge_mode, &cycles);
+				if (capture->pulse_capture) {
+					/* For PWM_CAPTURE_TYPE_PULSE */
+					capture->callback(dev, i, 0, cycles, status,
+							  capture->user_data);
+				} else {
+					/* For PWM_CAPTURE_TYPE_PERIOD */
+					capture->callback(dev, i, cycles, 0, status,
+							  capture->user_data);
+				}
+
+				if (capture->single_mode) {
+					EPWM_DisableCaptureInt(
+						epwm, i,
+						EPWM_CAPTURE_INT_RISING_LATCH |
+							EPWM_CAPTURE_INT_FALLING_LATCH);
+					data->capture[i].is_busy = false;
+				} else {
+					EPWM_ClearCaptureIntFlag(
+						epwm, i,
+						EPWM_CAPTURE_INT_FALLING_LATCH |
+							EPWM_CAPTURE_INT_RISING_LATCH);
+					EPWM_EnableCaptureInt(epwm, i,
+							      data->capture[i].curr_edge_mode);
+					/* data->capture[i].is_busy = true; */
+				}
+			}
+		}
+	}
+}
+
+static void pwm_numaker_p0_isr(const struct device *dev)
+{
+	/* Pair0 service channel 0, 1 */
+	__pwm_numaker_isr(dev, 0, 1);
+}
+
+static void pwm_numaker_p1_isr(const struct device *dev)
+{
+	/* Pair1 service channel 2, 3 */
+	__pwm_numaker_isr(dev, 2, 3);
+}
+
+static void pwm_numaker_p2_isr(const struct device *dev)
+{
+	/* Pair2 service channel 4, 5 */
+	__pwm_numaker_isr(dev, 4, 5);
+}
+#endif /* CONFIG_PWM_CAPTURE */
+
+/* PWM driver registration */
+static const struct pwm_driver_api pwm_numaker_driver_api = {
+	.set_cycles = pwm_numaker_set_cycles,
+	.get_cycles_per_sec = pwm_numaker_get_cycles_per_sec,
+#ifdef CONFIG_PWM_CAPTURE
+	.configure_capture = pwm_numaker_configure_capture,
+	.enable_capture = pwm_numaker_enable_capture,
+	.disable_capture = pwm_numaker_disable_capture,
+#endif /* CONFIG_PWM_CAPTURE */
+};
+
+/* Alternative EPWM clock get rate before support standard clock_control_get_rate */
+static int __pwm_numaker_clk_get_rate(EPWM_T *epwm, uint32_t *rate)
+{
+	uint32_t clk_src;
+	uint32_t epwm_clk_src;
+
+	if (epwm == EPWM0) {
+		clk_src = CLK->CLKSEL2 & CLK_CLKSEL2_EPWM0SEL_Msk;
+	} else { /* (epwm == EPWM1) */
```

Comment:
Would it be better to actually have that comment as code and in the `else` case a `LOG_ERR` and return error code? Would save you some debugging if you end up with a device with more than 2 channels in the future.